packages/firestore/src/local/indexeddb_remote_document_cache.ts

```diff
@@ -15,35 +15,71 @@
  */
 
 import { Query } from '../core/query';
-import { DocumentMap, documentMap } from '../model/collections';
-import { Document, MaybeDocument } from '../model/document';
+import {
+  documentKeySet,
+  DocumentKeySet,
+  DocumentMap,
+  documentMap
+} from '../model/collections';
+import { Document, MaybeDocument, NoDocument } from '../model/document';
 import { DocumentKey } from '../model/document_key';
 import { fail } from '../util/assert';
 
-import { DbRemoteDocument, DbRemoteDocumentKey } from './indexeddb_schema';
+import {
+  DbRemoteDocument,
+  DbRemoteDocumentKey,
+  DbSnapshotChange,
+  DbSnapshotChangeKey
+} from './indexeddb_schema';
 import { LocalSerializer } from './local_serializer';
 import { PersistenceTransaction } from './persistence';
 import { PersistencePromise } from './persistence_promise';
 import { RemoteDocumentCache } from './remote_document_cache';
 import { SimpleDbStore, SimpleDbTransaction } from './simple_db';
+import { SnapshotVersion } from '../core/snapshot_version';
+import { SortedMap } from '../util/sorted_map';
 
 export class IndexedDbRemoteDocumentCache implements RemoteDocumentCache {
   constructor(private serializer: LocalSerializer) {}
 
-  addEntry(
+  addEntries(
     transaction: PersistenceTransaction,
-    maybeDocument: MaybeDocument
+    maybeDocuments: MaybeDocument[]
   ): PersistencePromise<void> {
-    return remoteDocumentsStore(transaction).put(
-      dbKey(maybeDocument.key),
-      this.serializer.toDbRemoteDocument(maybeDocument)
+    const promises: Array<PersistencePromise<void>> = [];
+    const documentStore = remoteDocumentsStore(transaction);
+
+    let accumulatedChanges = new SortedMap<SnapshotVersion, DocumentKeySet>(
+      SnapshotVersion.comparator
     );
+
+    for (const maybeDocument of maybeDocuments) {
+      promises.push(
+        documentStore.put(
+          dbKey(maybeDocument.key),
+          this.serializer.toDbRemoteDocument(maybeDocument)
+        )
+      );
+
+      const existingChanges =
+        accumulatedChanges.get(maybeDocument.version) || documentKeySet();
```

Comment:
I am not sure we can use the versions of the documents as the key in the change log...  We can / will receive "old" documents (e.g. you do a new query that includes documents that haven't changed in months, etc.).

We could probably use the global snapshot that the document changes came from or something along those lines, but it's probably easier / sufficient to just use an IndexedDb autoIncrement key generator (`db.createObjectStore("names", { autoIncrement : true });` or whatever).  FWIW that means we could also ditch DbTimestampArray.